### PR TITLE
Flatten list type arrow chunks on parsing

### PR DESCRIPTION
**Pull request by arthurpassos**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
An issue with the following exception has been reported while trying to read a Parquet file from S3 into ClickHouse:

`['Code: 49.\nDB::Exception: offsets_column has data inconsistent with nested_column: While executing ParquetBlockInputFormat: While executing S3.`

This exception is thrown on [ColumnArray constructor](https://github.com/ClickHouse/ClickHouse/blob/master/src/Columns/ColumnArray.cpp#L59) whenever the offsets array doesn't match the data array.

The file in question has over 50 columns, but we managed to reproduce it with only two columns. One of the columns is of type Array(Int64) and the other Nullable(String). The file comes from Apache Iceberg. 

Reading these two columns with master's branch code, the returned arrow table contains two columns with two chunks each. After parsing it, the data column is duplicated, hence the mismatch with offsets column. The duplication seems to come from [getNestedArrowColumn](https://github.com/ClickHouse/ClickHouse/blob/master/src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp#L463) method. This method loops over the chunks of type `arrow::ListArray` and concatenates the return of `arrow::ListArray::values` without any sort of offsets. This assumes each chunk of type `arrow::ListArray` points to a different `arrow::ListArray::values_`, but that does not seem to be always true. It is not with the file I am testing.

Thus, it looks like the correct way to extract the nested values from an `arrow::ListArray` column is to slice based on the offsets of each chunk. **I'll gather evidence of this statement and update this description accordingly**.

The `arrow::ListArray` class offers [Flatten](https://github.com/apache/arrow/blob/master/cpp/src/arrow/array/array_nested.h#L149) method. The docs say: _Return an Array that is a concat of the lists in this array. Note that it's different from `values()` in that it takes into consideration of this array's offsets as well as null elements..._". [Flatten, under the hood, slices](https://github.com/apache/arrow/blob/master/cpp/src/arrow/array/array_nested.cc#L155) `arrow::ListArray::values` based on the offsets and length.

If the change in this [PR](https://github.com/ClickHouse/ClickHouse/pull/40485) is reverted, the returned table contains two columns with the Nullable(String) having two chunks and the Array(Int64) having only one single chunk. At first, I thought this was the issue, but as observed in [arrow batch reader example](https://github.com/apache/arrow/blob/master/cpp/src/parquet/arrow/arrow_reader_writer_test.cc#L3899), the layouts are expected to be different.

Unfortunately, the Parquet file I have is confidential and can't be included in this PR for testing. I have tried to anonymize the data in it, but once I do that, the error doesn't show up anymore. I have tried to generate several files using PySpark and arrow-cpp, but none of them caused the exception to happen. I did not have the time to look into how Apache Iceberg is generating these files.

I'll keep adding information to this PR as I gather it.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by arthurpassos** [comment]
@Avogar Can you enable CI and take a look?

**Comment by arthurpassos** [comment]
It looks like `arrow::ListArray::values()` (nested column data) might or might not be shared across chunks. When it is shared, the offsets will be monotonically increasing. Otherwise, the offsets will be zero based. This feels like an implementation detail and we are only facing issues with it because we are trying to access the underlying data structures to avoid unnecessary copies.

In order to properly handle both cases, I am proposing we slice the nested column data inconditionally and handle zero-based offsets by using a logic similar to the existing one, except that it also handles non-zero based.

There is an ongoing discussion about this with arrow maintainers in https://lists.apache.org/list?user@arrow.apache.org:2022-11.

**Comment by arthurpassos** [comment]
The Parquet test file attached contains a column of type List(Int64) divided into two chunks. Internally, arrow is representing this with two zero-based arrays.

The other scenario, in which the arrays would be monotonically increasing and the data would be shared across chunks, seems to only happen with a set of confidential files I have. As explained above, I failed to generate such test files. Therefore, for now, I haven't implemented any tests for this case.

File: tests/queries/0_stateless/02481_parquet_int_list_multiple_chunks.sh [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+#
+#arr = pa.array(gen_arrays(70000))
+#table  = pa.table([arr], ["arr"])
+#pq.write_table(table, "int-list-zero-based-chunked-array.parquet")
+
+DATA_FILE=$CUR_DIR/data_parquet/int-list-zero-based-chunked-array.parquet
+${CLICKHOUSE_CLIENT} --query="DROP TABLE IF EXISTS parquet_load"
+${CLICKHOUSE_CLIENT} --query="CREATE TABLE parquet_load (arr Array(Int64)) ENGINE = Memory"
+cat "$DATA_FILE" | ${CLICKHOUSE_CLIENT} -q "INSERT INTO parquet_load FORMAT Parquet"
+${CLICKHOUSE_CLIENT} --query="SELECT * FROM parquet_load"
```
Comment (Avogar) [inline tests/queries/0_stateless/02481_parquet_int_list_multiple_chunks.sh:40]: The reference file is too big. Let's use just checksum of this data
```suggestion
${CLICKHOUSE_CLIENT} --query="SELECT * FROM parquet_load" | md5sum
```

**Review by Avogar — APPROVED** [review]

**Comment by arthurpassos** [comment]
@Avogar super kind ping :)

**Comment by Avogar** [comment]
`AST fuzzer (debug)` - https://github.com/ClickHouse/ClickHouse/issues/14700
`Stateful tests` - known
`Stateless tests (debug)` - https://github.com/ClickHouse/ClickHouse/issues/43120
`Stress tests` - unrelated, but need to investigate

**Comment by den-crane** [comment]
Can it be backported? At least to 22.8?

**Pull request merged**